{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7bf79b80",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!TOKENIZERS_PARALLELISM=false"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7d175796",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p spacy_model_ner/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "758cdd09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: tabulate in /path/to/dir/.local/lib/python3.7/site-packages (0.8.9)\n",
      "\u001b[33mWARNING: There was an error checking the latest version of pip.\u001b[0m\u001b[33m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!pip install tabulate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "436e24f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import itertools\n",
    "import json\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ec4bc4a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import spacy\n",
    "nlp = spacy.load('en_core_web_sm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5f426ea9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spacy.require_gpu()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62ffdd6e",
   "metadata": {},
   "source": [
    "#### Load sentences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ca4766e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "UNLABELED_SENTENCE_DIR = '../data/diease_ner/unlabeled_sentences/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "eb0c0e3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sentences_df = pd.read_csv(f'{UNLABELED_SENTENCE_DIR}/complete_data_for_rules_tagging.csv',index_col=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8b01961d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Sentence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>Identification of APC2 , a homologue of the ad...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tr_1</td>\n",
       "      <td>The adenomatous polyposis coli ( APC ) tumour ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tr_10</td>\n",
       "      <td>A common MSH2 mutation in English and North Am...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tr_100</td>\n",
       "      <td>The positive control for DMT1 up - regulation ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tr_1000</td>\n",
       "      <td>The history further indicated intrauterine gro...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  New_Sentence_Id                                           Sentence\n",
       "0            tr_0  Identification of APC2 , a homologue of the ad...\n",
       "1            tr_1  The adenomatous polyposis coli ( APC ) tumour ...\n",
       "2           tr_10  A common MSH2 mutation in English and North Am...\n",
       "3          tr_100  The positive control for DMT1 up - regulation ...\n",
       "4         tr_1000  The history further indicated intrauterine gro..."
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentences_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "862c7503",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Identification of APC2 , a homologue of the adenomatous polyposis coli tumour suppressor .'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentences_df.loc[0,'Sentence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ca22f836",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A common MSH2 mutation in English and North American HNPCC families : origin , phenotypic expression , and sex specific differences in colorectal cancer .'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentences_df.loc[2,'Sentence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c7f1c1de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7298, 2)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentences_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "01b8bb90",
   "metadata": {},
   "outputs": [],
   "source": [
    "text = sentences_df.loc[0,'Sentence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cc0c4a77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Identification of APC2 , a homologue of the adenomatous polyposis coli tumour suppressor .'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9f65b3af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Identification', 'NN', 'nsubj'),\n",
       " ('of', 'IN', 'prep'),\n",
       " ('APC2', 'NNP', 'pobj'),\n",
       " (',', ',', 'punct'),\n",
       " ('a', 'DT', 'det'),\n",
       " ('homologue', 'NN', 'appos'),\n",
       " ('of', 'IN', 'prep'),\n",
       " ('the', 'DT', 'det'),\n",
       " ('adenomatous', 'JJ', 'amod'),\n",
       " ('polyposis', 'NN', 'pobj'),\n",
       " ('coli', 'VBZ', 'ROOT'),\n",
       " ('tumour', 'NN', 'compound'),\n",
       " ('suppressor', 'NN', 'dobj'),\n",
       " ('.', '.', 'punct')]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doc = nlp(text)\n",
    "\n",
    "[(token.text, token.tag_, token.dep_) for token in doc]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c14ab590",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p ../data/diease_ner/diseases_words/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a982b920",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/diease_ner/diseases_words/diseases_ner.json','r') as f:\n",
    "    diseases_json = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "effa6b7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "diseases_ner.json\r\n",
      "diseases_ner.sjon\r\n",
      "split_of_classes_sentence_level.csv\r\n",
      "split_of_classes_sentence_level_dev.csv\r\n",
      "split_of_classes_sentence_level_test.csv\r\n",
      "split_of_classes_sentence_level_train.csv\r\n",
      "token_level_tags_on_one_unlabeled_sentence.csv\r\n",
      "token_level_tags_on_one_unlabeled_sentence_2.csv\r\n"
     ]
    }
   ],
   "source": [
    "!ls spacy_model_ner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a316d5a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp ../data/diease_ner/diseases_words/diseases_ner.json spacy_model_ner/diseases_ner.json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28a3c67e",
   "metadata": {},
   "source": [
    "Source for the disease words: <br>\n",
    "> thanks to Shivanshu Gupta to have crawled the below source of diseases. We can use this as a base list to produce our rules patterns\n",
    "> https://raw.githubusercontent.com/Shivanshu-Gupta/web-scrapers/master/medical_ner/medicinenet-diseases.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6463ffa4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'disease': 'Hemophilia'},\n",
       " {'disease': 'A, Hemophilia'},\n",
       " {'disease': 'Hemophilia A'},\n",
       " {'disease': 'Hepatitis A'},\n",
       " {'disease': 'A, Hepatitis'}]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diseases_json[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0a6a2d41",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_diseases = [each['disease'] for each in diseases_json if not re.search('[,]|test',each['disease'],re.I)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fcee3401",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4969\n"
     ]
    }
   ],
   "source": [
    "print(len(diseases_json))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "fd585b58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4283\n"
     ]
    }
   ],
   "source": [
    "print(len(list_of_diseases))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "838bbcda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Hemophilia',\n",
       " 'Hemophilia A',\n",
       " 'Hepatitis A',\n",
       " 'Abdominal Aortic Aneurysm',\n",
       " 'AAA']"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_diseases[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "0cecc7a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_diseases.extend(['tumor','tumour'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "8a1f79d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def list_of_words_2_spacy_patterns(list_of_words,\n",
    "                                   nlp_model,\n",
    "                                   label_name\n",
    "                                  ):\n",
    "    spacy_patterns = []\n",
    "    for each_word in list_of_words:\n",
    "        sub_pattern_list = [] # [{\"ORTH\": user_text_entity_df.loc[each_pattern_index,'TEXT']}]\n",
    "        for token in  nlp_model(each_word.lower()):\n",
    "            if re.search('^\\W{1,}$',token.text):\n",
    "                sub_pattern_list.append({\"ORTH\": token.text,\"OP\":\"*\"})\n",
    "            else:\n",
    "                sub_pattern_list.append({\"LOWER\":token.text})\n",
    "        temp_dict = {\"label\": label_name,\n",
    "                     \"pattern\": sub_pattern_list}\n",
    "        spacy_patterns.append(temp_dict)\n",
    "    return spacy_patterns\n",
    "\n",
    "\n",
    "def load_rules_nlp_model_from_spacy_patterns(spacy_patterns):\n",
    "    rules_nlp = spacy.load('en_core_web_sm',disable=['ner'])\n",
    "    rules_config = {\n",
    "        \"validate\": True,\n",
    "        \"overwrite_ents\": True,\n",
    "    }\n",
    "\n",
    "    disease_rules = rules_nlp.add_pipe(\"entity_ruler\", # invoke entity_ruler pipe \n",
    "                                       \"disease_rules\", # give a name to the pipe\n",
    "                                       config=rules_config)\n",
    "    disease_rules.add_patterns(spacy_patterns)\n",
    "    return rules_nlp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "018bc408",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_spacy_rules_patterns = list_of_words_2_spacy_patterns(list_of_diseases,\n",
    "                               nlp,\n",
    "                               \"DISEASE\"\n",
    "                              )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e37fdc11",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_ner_rules_nlp = load_rules_nlp_model_from_spacy_patterns(disease_spacy_rules_patterns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "10659d27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['tok2vec',\n",
       " 'tagger',\n",
       " 'parser',\n",
       " 'attribute_ruler',\n",
       " 'lemmatizer',\n",
       " 'disease_rules']"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_ner_rules_nlp.pipe_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "87fa2559",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DISEASE']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_ner_rules_nlp.pipe_labels['disease_rules']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "186fe7ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p ../data/model_weights/spacy_rules\n",
    "!mkdir -p ../data/model_weights/spacy_rules_packaged_model/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d113a95a",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_ner_rules_nlp.to_disk('../data/model_weights/spacy_rules')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "feb2e8d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "conll_compatible_ner_data\tdiease_ner    ner_sentences_level_data\r\n",
      "conll_spacy_tokenized_ner_data\tmodel_config\r\n"
     ]
    }
   ],
   "source": [
    "!ls ../data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "94419017",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[38;5;4mℹ Building package artifacts: sdist\u001b[0m\n",
      "\u001b[38;5;2m✔ Loaded meta.json from file\u001b[0m\n",
      "../data/model_weights/spacy_rules/meta.json\n",
      "\u001b[38;5;2m✔ Generated README.md from meta.json\u001b[0m\n",
      "\u001b[38;5;2m✔ Successfully created package directory\n",
      "'en_disease_ner_rules-3.3.0'\u001b[0m\n",
      "../data/model_weights/spacy_rules_packaged_model/en_disease_ner_rules-3.3.0\n",
      "running sdist\n",
      "running egg_info\n",
      "creating en_disease_ner_rules.egg-info\n",
      "writing en_disease_ner_rules.egg-info/PKG-INFO\n",
      "writing dependency_links to en_disease_ner_rules.egg-info/dependency_links.txt\n",
      "writing entry points to en_disease_ner_rules.egg-info/entry_points.txt\n",
      "writing requirements to en_disease_ner_rules.egg-info/requires.txt\n",
      "writing top-level names to en_disease_ner_rules.egg-info/top_level.txt\n",
      "writing manifest file 'en_disease_ner_rules.egg-info/SOURCES.txt'\n",
      "reading manifest file 'en_disease_ner_rules.egg-info/SOURCES.txt'\n",
      "reading manifest template 'MANIFEST.in'\n",
      "warning: no files found matching 'LICENSE'\n",
      "warning: no files found matching 'LICENSES_SOURCES'\n",
      "writing manifest file 'en_disease_ner_rules.egg-info/SOURCES.txt'\n",
      "running check\n",
      "creating en_disease_ner_rules-3.3.0\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules.egg-info\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/attribute_ruler\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/disease_rules\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/lemmatizer\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/lemmatizer/lookups\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/ner\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/parser\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/senter\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/tagger\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/tok2vec\n",
      "creating en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/vocab\n",
      "copying files to en_disease_ner_rules-3.3.0...\n",
      "copying MANIFEST.in -> en_disease_ner_rules-3.3.0\n",
      "copying README.md -> en_disease_ner_rules-3.3.0\n",
      "copying meta.json -> en_disease_ner_rules-3.3.0\n",
      "copying setup.py -> en_disease_ner_rules-3.3.0\n",
      "copying en_disease_ner_rules/__init__.py -> en_disease_ner_rules-3.3.0/en_disease_ner_rules\n",
      "copying en_disease_ner_rules/meta.json -> en_disease_ner_rules-3.3.0/en_disease_ner_rules\n",
      "copying en_disease_ner_rules.egg-info/PKG-INFO -> en_disease_ner_rules-3.3.0/en_disease_ner_rules.egg-info\n",
      "copying en_disease_ner_rules.egg-info/SOURCES.txt -> en_disease_ner_rules-3.3.0/en_disease_ner_rules.egg-info\n",
      "copying en_disease_ner_rules.egg-info/dependency_links.txt -> en_disease_ner_rules-3.3.0/en_disease_ner_rules.egg-info\n",
      "copying en_disease_ner_rules.egg-info/entry_points.txt -> en_disease_ner_rules-3.3.0/en_disease_ner_rules.egg-info\n",
      "copying en_disease_ner_rules.egg-info/not-zip-safe -> en_disease_ner_rules-3.3.0/en_disease_ner_rules.egg-info\n",
      "copying en_disease_ner_rules.egg-info/requires.txt -> en_disease_ner_rules-3.3.0/en_disease_ner_rules.egg-info\n",
      "copying en_disease_ner_rules.egg-info/top_level.txt -> en_disease_ner_rules-3.3.0/en_disease_ner_rules.egg-info\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/README.md -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/config.cfg -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/meta.json -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/tokenizer -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/attribute_ruler/patterns -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/attribute_ruler\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/disease_rules/cfg -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/disease_rules\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/disease_rules/patterns.jsonl -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/disease_rules\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/lemmatizer/lookups/lookups.bin -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/lemmatizer/lookups\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/ner/cfg -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/ner\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/ner/model -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/ner\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/ner/moves -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/ner\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/parser/cfg -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/parser\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/parser/model -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/parser\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/parser/moves -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/parser\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/senter/cfg -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/senter\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/senter/model -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/senter\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/tagger/cfg -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/tagger\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/tagger/model -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/tagger\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/tok2vec/cfg -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/tok2vec\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/tok2vec/model -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/tok2vec\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/vocab/key2row -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/vocab\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/vocab/lookups.bin -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/vocab\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/vocab/strings.json -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/vocab\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/vocab/vectors -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/vocab\n",
      "copying en_disease_ner_rules/en_disease_ner_rules-3.3.0/vocab/vectors.cfg -> en_disease_ner_rules-3.3.0/en_disease_ner_rules/en_disease_ner_rules-3.3.0/vocab\n",
      "Writing en_disease_ner_rules-3.3.0/setup.cfg\n",
      "creating dist\n",
      "Creating tar archive\n",
      "removing 'en_disease_ner_rules-3.3.0' (and everything under it)\n",
      "\u001b[38;5;2m✔ Successfully created zipped Python package\u001b[0m\n",
      "../data/model_weights/spacy_rules_packaged_model/en_disease_ner_rules-3.3.0/dist/en_disease_ner_rules-3.3.0.tar.gz\n"
     ]
    }
   ],
   "source": [
    "!python -m spacy package ../data/model_weights/spacy_rules/ ../data/model_weights/spacy_rules_packaged_model/ --name disease_ner_rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "563b9e9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Sentence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7293</th>\n",
       "      <td>tes_95</td>\n",
       "      <td>Deletion mapping studies have unambiguously id...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7294</th>\n",
       "      <td>tes_96</td>\n",
       "      <td>A new tumor suppressor gene , PTEN / MMAC1 , w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7295</th>\n",
       "      <td>tes_97</td>\n",
       "      <td>We screened 80 prostate tumors by microsatelli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7296</th>\n",
       "      <td>tes_98</td>\n",
       "      <td>We then proceeded with sequence analysis of th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7297</th>\n",
       "      <td>tes_99</td>\n",
       "      <td>The identification of the second mutational ev...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     New_Sentence_Id                                           Sentence\n",
       "7293          tes_95  Deletion mapping studies have unambiguously id...\n",
       "7294          tes_96  A new tumor suppressor gene , PTEN / MMAC1 , w...\n",
       "7295          tes_97  We screened 80 prostate tumors by microsatelli...\n",
       "7296          tes_98  We then proceeded with sequence analysis of th...\n",
       "7297          tes_99  The identification of the second mutational ev..."
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentences_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "c4a99f7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7298, 2)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentences_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "bfddaf88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "New_Sentence_Id    0\n",
       "Sentence           3\n",
       "dtype: int64"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentences_df.isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "9b6df2f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sentences_df.dropna(subset=['Sentence'],axis=0,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "efdd1e1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "New_Sentence_Id    0\n",
       "Sentence           0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentences_df.reset_index(drop=True,inplace=True)\n",
    "sentences_df.isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "7374c4ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7295, 2)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentences_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39e4056d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e0125f9c",
   "metadata": {},
   "source": [
    "### Prepare rules output NER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "26a27897",
   "metadata": {},
   "outputs": [],
   "source": [
    "sentences_ids_tuples = list(zip(sentences_df['Sentence'],sentences_df['New_Sentence_Id']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "e98fc5fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process 0 sentences\n",
      "Process 100 sentences\n",
      "Process 200 sentences\n",
      "Process 300 sentences\n",
      "Process 400 sentences\n",
      "Process 500 sentences\n",
      "Process 600 sentences\n",
      "Process 700 sentences\n",
      "Process 800 sentences\n",
      "Process 900 sentences\n",
      "Process 1000 sentences\n",
      "Process 1100 sentences\n",
      "Process 1200 sentences\n",
      "Process 1300 sentences\n",
      "Process 1400 sentences\n",
      "Process 1500 sentences\n",
      "Process 1600 sentences\n",
      "Process 1700 sentences\n",
      "Process 1800 sentences\n",
      "Process 1900 sentences\n",
      "Process 2000 sentences\n",
      "Process 2100 sentences\n",
      "Process 2200 sentences\n",
      "Process 2300 sentences\n",
      "Process 2400 sentences\n",
      "Process 2500 sentences\n",
      "Process 2600 sentences\n",
      "Process 2700 sentences\n",
      "Process 2800 sentences\n",
      "Process 2900 sentences\n",
      "Process 3000 sentences\n",
      "Process 3100 sentences\n",
      "Process 3200 sentences\n",
      "Process 3300 sentences\n",
      "Process 3400 sentences\n",
      "Process 3500 sentences\n",
      "Process 3600 sentences\n",
      "Process 3700 sentences\n",
      "Process 3800 sentences\n",
      "Process 3900 sentences\n",
      "Process 4000 sentences\n",
      "Process 4100 sentences\n",
      "Process 4200 sentences\n",
      "Process 4300 sentences\n",
      "Process 4400 sentences\n",
      "Process 4500 sentences\n",
      "Process 4600 sentences\n",
      "Process 4700 sentences\n",
      "Process 4800 sentences\n",
      "Process 4900 sentences\n",
      "Process 5000 sentences\n",
      "Process 5100 sentences\n",
      "Process 5200 sentences\n",
      "Process 5300 sentences\n",
      "Process 5400 sentences\n",
      "Process 5500 sentences\n",
      "Process 5600 sentences\n",
      "Process 5700 sentences\n",
      "Process 5800 sentences\n",
      "Process 5900 sentences\n",
      "Process 6000 sentences\n",
      "Process 6100 sentences\n",
      "Process 6200 sentences\n",
      "Process 6300 sentences\n",
      "Process 6400 sentences\n",
      "Process 6500 sentences\n",
      "Process 6600 sentences\n",
      "Process 6700 sentences\n",
      "Process 6800 sentences\n",
      "Process 6900 sentences\n",
      "Process 7000 sentences\n",
      "Process 7100 sentences\n",
      "Process 7200 sentences\n",
      "CPU times: user 20.4 s, sys: 186 ms, total: 20.5 s\n",
      "Wall time: 20.5 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "token_tag_tuples = []\n",
    "for i, (doc, doc_id) in enumerate(disease_ner_rules_nlp.pipe(sentences_ids_tuples,\n",
    "                                                           as_tuples=True,\n",
    "                                                           batch_size=100,\n",
    "                                                          )                                \n",
    "                               ):    \n",
    "    if i%100 == 0:\n",
    "        print(f\"Process {i} sentences\")\n",
    "    for token in doc:\n",
    "        token_ent_type = 'O' if token.ent_type_ == '' else token.ent_type_\n",
    "        token_tag_tuples.append((doc_id,token.text,token_ent_type))\n",
    "    token_tag_tuples.append((doc_id, '[SEP]', '[SEP]'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "41bea35f",
   "metadata": {},
   "outputs": [],
   "source": [
    "token_tag_df = pd.DataFrame(token_tag_tuples,columns=['New_Sentence_Id','Token','Rules_Tag'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "269b37c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_token_df_2_conll_format(token_tag_df,\n",
    "                                    token_column_name,\n",
    "                                    tag_column_name\n",
    "                                   ):\n",
    "    new_tag_column_name = tag_column_name + '_BIO'\n",
    "    token_tag_df[new_tag_column_name] = ''\n",
    "    # token_string = ''\n",
    "    previous_tag_string = 'O'\n",
    "    for each in range(len(token_tag_df)):\n",
    "        if each %1000 == 0:\n",
    "            print(f\"{each} tokens processed\")\n",
    "        current_token_string = str(token_tag_df.loc[each,token_column_name])\n",
    "        current_tag_string = str(token_tag_df.loc[each,tag_column_name])\n",
    "        \n",
    "        if current_token_string !='[SEP]':\n",
    "            if previous_tag_string == 'O':\n",
    "                if current_tag_string != 'O':\n",
    "                    current_line = current_token_string + \"\\t\" + \"B-\" + current_tag_string + \"\\n\"\n",
    "                else:\n",
    "                    current_line = current_token_string + \"\\t\" + current_tag_string + \"\\n\"\n",
    "            else:\n",
    "                if current_tag_string == previous_tag_string:\n",
    "                    current_line = current_token_string + \"\\t\" + \"I-\" + current_tag_string + \"\\n\"\n",
    "                else:\n",
    "                    if current_tag_string != 'O':\n",
    "                        current_line = current_token_string + \"\\t\" + \"B-\" + current_tag_string + \"\\n\"\n",
    "                    else:\n",
    "                        current_line = current_token_string + \"\\t\" + current_tag_string + \"\\n\"\n",
    "                        \n",
    "            token_tag_df.loc[each,new_tag_column_name] = current_line.strip().split(\"\\t\")[1]\n",
    "        else:\n",
    "            current_line = \"\\n\"\n",
    "            token_tag_df.loc[each,new_tag_column_name] = '[SEP]'\n",
    "        # token_string = token_string + current_line\n",
    "        previous_tag_string = current_tag_string\n",
    "        \n",
    "    return token_tag_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "68200fe6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 tokens processed\n",
      "1000 tokens processed\n",
      "2000 tokens processed\n",
      "3000 tokens processed\n",
      "4000 tokens processed\n",
      "5000 tokens processed\n",
      "6000 tokens processed\n",
      "7000 tokens processed\n",
      "8000 tokens processed\n",
      "9000 tokens processed\n",
      "10000 tokens processed\n",
      "11000 tokens processed\n",
      "12000 tokens processed\n",
      "13000 tokens processed\n",
      "14000 tokens processed\n",
      "15000 tokens processed\n",
      "16000 tokens processed\n",
      "17000 tokens processed\n",
      "18000 tokens processed\n",
      "19000 tokens processed\n",
      "20000 tokens processed\n",
      "21000 tokens processed\n",
      "22000 tokens processed\n",
      "23000 tokens processed\n",
      "24000 tokens processed\n",
      "25000 tokens processed\n",
      "26000 tokens processed\n",
      "27000 tokens processed\n",
      "28000 tokens processed\n",
      "29000 tokens processed\n",
      "30000 tokens processed\n",
      "31000 tokens processed\n",
      "32000 tokens processed\n",
      "33000 tokens processed\n",
      "34000 tokens processed\n",
      "35000 tokens processed\n",
      "36000 tokens processed\n",
      "37000 tokens processed\n",
      "38000 tokens processed\n",
      "39000 tokens processed\n",
      "40000 tokens processed\n",
      "41000 tokens processed\n",
      "42000 tokens processed\n",
      "43000 tokens processed\n",
      "44000 tokens processed\n",
      "45000 tokens processed\n",
      "46000 tokens processed\n",
      "47000 tokens processed\n",
      "48000 tokens processed\n",
      "49000 tokens processed\n",
      "50000 tokens processed\n",
      "51000 tokens processed\n",
      "52000 tokens processed\n",
      "53000 tokens processed\n",
      "54000 tokens processed\n",
      "55000 tokens processed\n",
      "56000 tokens processed\n",
      "57000 tokens processed\n",
      "58000 tokens processed\n",
      "59000 tokens processed\n",
      "60000 tokens processed\n",
      "61000 tokens processed\n",
      "62000 tokens processed\n",
      "63000 tokens processed\n",
      "64000 tokens processed\n",
      "65000 tokens processed\n",
      "66000 tokens processed\n",
      "67000 tokens processed\n",
      "68000 tokens processed\n",
      "69000 tokens processed\n",
      "70000 tokens processed\n",
      "71000 tokens processed\n",
      "72000 tokens processed\n",
      "73000 tokens processed\n",
      "74000 tokens processed\n",
      "75000 tokens processed\n",
      "76000 tokens processed\n",
      "77000 tokens processed\n",
      "78000 tokens processed\n",
      "79000 tokens processed\n",
      "80000 tokens processed\n",
      "81000 tokens processed\n",
      "82000 tokens processed\n",
      "83000 tokens processed\n",
      "84000 tokens processed\n",
      "85000 tokens processed\n",
      "86000 tokens processed\n",
      "87000 tokens processed\n",
      "88000 tokens processed\n",
      "89000 tokens processed\n",
      "90000 tokens processed\n",
      "91000 tokens processed\n",
      "92000 tokens processed\n",
      "93000 tokens processed\n",
      "94000 tokens processed\n",
      "95000 tokens processed\n",
      "96000 tokens processed\n",
      "97000 tokens processed\n",
      "98000 tokens processed\n",
      "99000 tokens processed\n",
      "100000 tokens processed\n",
      "101000 tokens processed\n",
      "102000 tokens processed\n",
      "103000 tokens processed\n",
      "104000 tokens processed\n",
      "105000 tokens processed\n",
      "106000 tokens processed\n",
      "107000 tokens processed\n",
      "108000 tokens processed\n",
      "109000 tokens processed\n",
      "110000 tokens processed\n",
      "111000 tokens processed\n",
      "112000 tokens processed\n",
      "113000 tokens processed\n",
      "114000 tokens processed\n",
      "115000 tokens processed\n",
      "116000 tokens processed\n",
      "117000 tokens processed\n",
      "118000 tokens processed\n",
      "119000 tokens processed\n",
      "120000 tokens processed\n",
      "121000 tokens processed\n",
      "122000 tokens processed\n",
      "123000 tokens processed\n",
      "124000 tokens processed\n",
      "125000 tokens processed\n",
      "126000 tokens processed\n",
      "127000 tokens processed\n",
      "128000 tokens processed\n",
      "129000 tokens processed\n",
      "130000 tokens processed\n",
      "131000 tokens processed\n",
      "132000 tokens processed\n",
      "133000 tokens processed\n",
      "134000 tokens processed\n",
      "135000 tokens processed\n",
      "136000 tokens processed\n",
      "137000 tokens processed\n",
      "138000 tokens processed\n",
      "139000 tokens processed\n",
      "140000 tokens processed\n",
      "141000 tokens processed\n",
      "142000 tokens processed\n",
      "143000 tokens processed\n",
      "144000 tokens processed\n",
      "145000 tokens processed\n",
      "146000 tokens processed\n",
      "147000 tokens processed\n",
      "148000 tokens processed\n",
      "149000 tokens processed\n",
      "150000 tokens processed\n",
      "151000 tokens processed\n",
      "152000 tokens processed\n",
      "153000 tokens processed\n",
      "154000 tokens processed\n",
      "155000 tokens processed\n",
      "156000 tokens processed\n",
      "157000 tokens processed\n",
      "158000 tokens processed\n",
      "159000 tokens processed\n",
      "160000 tokens processed\n",
      "161000 tokens processed\n",
      "162000 tokens processed\n",
      "163000 tokens processed\n",
      "164000 tokens processed\n",
      "165000 tokens processed\n",
      "166000 tokens processed\n",
      "167000 tokens processed\n",
      "168000 tokens processed\n",
      "169000 tokens processed\n",
      "170000 tokens processed\n",
      "171000 tokens processed\n",
      "172000 tokens processed\n",
      "173000 tokens processed\n",
      "174000 tokens processed\n",
      "175000 tokens processed\n",
      "176000 tokens processed\n",
      "177000 tokens processed\n",
      "178000 tokens processed\n",
      "179000 tokens processed\n",
      "180000 tokens processed\n",
      "181000 tokens processed\n",
      "182000 tokens processed\n",
      "183000 tokens processed\n",
      "184000 tokens processed\n",
      "185000 tokens processed\n",
      "186000 tokens processed\n",
      "187000 tokens processed\n",
      "188000 tokens processed\n",
      "189000 tokens processed\n",
      "190000 tokens processed\n",
      "191000 tokens processed\n",
      "CPU times: user 9min 1s, sys: 331 ms, total: 9min 1s\n",
      "Wall time: 9min 1s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "token_tag_df_2 = convert_token_df_2_conll_format(token_tag_df,\n",
    "                                                    'Token',\n",
    "                                                    'Rules_Tag'\n",
    "                                                   )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "b1ff1da8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>Identification</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>of</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>APC2</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>a</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  New_Sentence_Id           Token Rules_Tag Rules_Tag_BIO\n",
       "0            tr_0  Identification         O             O\n",
       "1            tr_0              of         O             O\n",
       "2            tr_0            APC2         O             O\n",
       "3            tr_0               ,         O             O\n",
       "4            tr_0               a         O             O"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "token_tag_df_2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "01d42f53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n"
     ]
    }
   ],
   "source": [
    "SPACY_RULES_NER_DIR = '../data/diease_ner/rules_ner'\n",
    "!mkdir -p $SPACY_RULES_NER_DIR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "cd357084",
   "metadata": {},
   "outputs": [],
   "source": [
    "token_tag_df_2.to_csv(f'{SPACY_RULES_NER_DIR}/token_level_spacy_rules_ner_output.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5a9ff2bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>Identification</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>of</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>APC2</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>a</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>homologue</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>of</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>the</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>adenomatous</td>\n",
       "      <td>DISEASE</td>\n",
       "      <td>B-DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>polyposis</td>\n",
       "      <td>DISEASE</td>\n",
       "      <td>I-DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>coli</td>\n",
       "      <td>DISEASE</td>\n",
       "      <td>I-DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>tumour</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>suppressor</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>.</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>tr_1</td>\n",
       "      <td>The</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>tr_1</td>\n",
       "      <td>adenomatous</td>\n",
       "      <td>DISEASE</td>\n",
       "      <td>B-DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>tr_1</td>\n",
       "      <td>polyposis</td>\n",
       "      <td>DISEASE</td>\n",
       "      <td>I-DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>tr_1</td>\n",
       "      <td>coli</td>\n",
       "      <td>DISEASE</td>\n",
       "      <td>I-DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tr_1</td>\n",
       "      <td>(</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>tr_1</td>\n",
       "      <td>APC</td>\n",
       "      <td>DISEASE</td>\n",
       "      <td>B-DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>tr_1</td>\n",
       "      <td>)</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>tr_1</td>\n",
       "      <td>tumour</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>tr_1</td>\n",
       "      <td>-</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>tr_1</td>\n",
       "      <td>suppressor</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   New_Sentence_Id           Token Rules_Tag Rules_Tag_BIO\n",
       "0             tr_0  Identification         O             O\n",
       "1             tr_0              of         O             O\n",
       "2             tr_0            APC2         O             O\n",
       "3             tr_0               ,         O             O\n",
       "4             tr_0               a         O             O\n",
       "5             tr_0       homologue         O             O\n",
       "6             tr_0              of         O             O\n",
       "7             tr_0             the         O             O\n",
       "8             tr_0     adenomatous   DISEASE     B-DISEASE\n",
       "9             tr_0       polyposis   DISEASE     I-DISEASE\n",
       "10            tr_0            coli   DISEASE     I-DISEASE\n",
       "11            tr_0          tumour         O             O\n",
       "12            tr_0      suppressor         O             O\n",
       "13            tr_0               .         O             O\n",
       "14            tr_0           [SEP]     [SEP]         [SEP]\n",
       "15            tr_1             The         O             O\n",
       "16            tr_1     adenomatous   DISEASE     B-DISEASE\n",
       "17            tr_1       polyposis   DISEASE     I-DISEASE\n",
       "18            tr_1            coli   DISEASE     I-DISEASE\n",
       "19            tr_1               (         O             O\n",
       "20            tr_1             APC   DISEASE     B-DISEASE\n",
       "21            tr_1               )         O             O\n",
       "22            tr_1          tumour         O             O\n",
       "23            tr_1               -         O             O\n",
       "24            tr_1      suppressor         O             O"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "token_tag_df_2.head(25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c4d5dde",
   "metadata": {},
   "source": [
    "### Adding human annotated tags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "7115d05e",
   "metadata": {},
   "outputs": [],
   "source": [
    "SPACY_RULES_NER_DIR = '../data/diease_ner/rules_ner'\n",
    "\n",
    "token_tag_df_2 = pd.read_csv(f'{SPACY_RULES_NER_DIR}/token_level_spacy_rules_ner_output.csv',index_col=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "a9513802",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>Chromosomal</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>analysis</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>was</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>normal</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>(</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>46</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>XY</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>)</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>.</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    New_Sentence_Id        Token Rules_Tag Rules_Tag_BIO\n",
       "205         tr_1007  Chromosomal         O             O\n",
       "206         tr_1007     analysis         O             O\n",
       "207         tr_1007          was         O             O\n",
       "208         tr_1007       normal         O             O\n",
       "209         tr_1007            (         O             O\n",
       "210         tr_1007           46         O             O\n",
       "211         tr_1007            ,         O             O\n",
       "212         tr_1007           XY         O             O\n",
       "213         tr_1007            )         O             O\n",
       "214         tr_1007            .         O             O\n",
       "215         tr_1007        [SEP]     [SEP]         [SEP]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "token_tag_df_2[token_tag_df_2['New_Sentence_Id'] == 'tr_1007']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "5777605d",
   "metadata": {},
   "outputs": [],
   "source": [
    "SPACY_TOKENIZED_DIR = '../data/diease_ner/conll_spacy_tokenized_ner_data/'\n",
    "final_token_level_annotated_data = pd.read_csv(f'{SPACY_TOKENIZED_DIR}/token_level_annotated_data.csv',index_col=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "035b92f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Spacy_Small_Tokens</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26281</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>Chromosomal</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26282</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>analysis</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26283</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>was</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26284</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>normal</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26285</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>(</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26286</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>46</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26287</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26288</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>XY</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26289</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>)</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26290</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>.</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26291</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      New_Sentence_Id Spacy_Small_Tokens Human_Annotated_Tag_BIO\n",
       "26281         tr_1007        Chromosomal                       O\n",
       "26282         tr_1007           analysis                       O\n",
       "26283         tr_1007                was                       O\n",
       "26284         tr_1007             normal                       O\n",
       "26285         tr_1007                  (                       O\n",
       "26286         tr_1007                 46                       O\n",
       "26287         tr_1007                  ,                       O\n",
       "26288         tr_1007                 XY                       O\n",
       "26289         tr_1007                  )                       O\n",
       "26290         tr_1007                  .                       O\n",
       "26291         tr_1007              [SEP]                   [SEP]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_token_level_annotated_data[final_token_level_annotated_data['New_Sentence_Id'] == 'tr_1007']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "67c500b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_token_level_annotated_data.columns = ['New_Sentence_Id', 'Token','Human_Annotated_Tag_BIO']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "0be916a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_token_level_annotated_data['New_index'] = [every for every in range(len(final_token_level_annotated_data))]\n",
    "final_token_level_annotated_data['New_index'] = final_token_level_annotated_data['New_index'].astype(str)\n",
    "final_token_level_annotated_data['New_Spacy_Tokens'] = final_token_level_annotated_data['New_index'] + final_token_level_annotated_data['Token']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "c5f4d3d2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "      <th>New_index</th>\n",
       "      <th>New_Spacy_Tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>Identification</td>\n",
       "      <td>O</td>\n",
       "      <td>0</td>\n",
       "      <td>0Identification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>of</td>\n",
       "      <td>O</td>\n",
       "      <td>1</td>\n",
       "      <td>1of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>APC2</td>\n",
       "      <td>O</td>\n",
       "      <td>2</td>\n",
       "      <td>2APC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>3</td>\n",
       "      <td>3,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>a</td>\n",
       "      <td>O</td>\n",
       "      <td>4</td>\n",
       "      <td>4a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  New_Sentence_Id           Token Human_Annotated_Tag_BIO New_index  \\\n",
       "0            tr_0  Identification                       O         0   \n",
       "1            tr_0              of                       O         1   \n",
       "2            tr_0            APC2                       O         2   \n",
       "3            tr_0               ,                       O         3   \n",
       "4            tr_0               a                       O         4   \n",
       "\n",
       "  New_Spacy_Tokens  \n",
       "0  0Identification  \n",
       "1              1of  \n",
       "2            2APC2  \n",
       "3               3,  \n",
       "4               4a  "
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_token_level_annotated_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "5b1e1adc",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_token_level_annotated_data.sort_values(by=['New_Sentence_Id','New_Spacy_Tokens'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "007afa11",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_token_level_annotated_data.reset_index(drop=True,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "438c3374",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "      <th>New_index</th>\n",
       "      <th>New_Spacy_Tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25685</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>Chromosomal</td>\n",
       "      <td>O</td>\n",
       "      <td>26281</td>\n",
       "      <td>26281Chromosomal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25686</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>analysis</td>\n",
       "      <td>O</td>\n",
       "      <td>26282</td>\n",
       "      <td>26282analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25687</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>was</td>\n",
       "      <td>O</td>\n",
       "      <td>26283</td>\n",
       "      <td>26283was</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25688</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>normal</td>\n",
       "      <td>O</td>\n",
       "      <td>26284</td>\n",
       "      <td>26284normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25689</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>(</td>\n",
       "      <td>O</td>\n",
       "      <td>26285</td>\n",
       "      <td>26285(</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25690</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>46</td>\n",
       "      <td>O</td>\n",
       "      <td>26286</td>\n",
       "      <td>2628646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25691</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>26287</td>\n",
       "      <td>26287,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25692</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>XY</td>\n",
       "      <td>O</td>\n",
       "      <td>26288</td>\n",
       "      <td>26288XY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25693</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>)</td>\n",
       "      <td>O</td>\n",
       "      <td>26289</td>\n",
       "      <td>26289)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25694</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>.</td>\n",
       "      <td>O</td>\n",
       "      <td>26290</td>\n",
       "      <td>26290.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25695</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>26291</td>\n",
       "      <td>26291[SEP]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      New_Sentence_Id        Token Human_Annotated_Tag_BIO New_index  \\\n",
       "25685         tr_1007  Chromosomal                       O     26281   \n",
       "25686         tr_1007     analysis                       O     26282   \n",
       "25687         tr_1007          was                       O     26283   \n",
       "25688         tr_1007       normal                       O     26284   \n",
       "25689         tr_1007            (                       O     26285   \n",
       "25690         tr_1007           46                       O     26286   \n",
       "25691         tr_1007            ,                       O     26287   \n",
       "25692         tr_1007           XY                       O     26288   \n",
       "25693         tr_1007            )                       O     26289   \n",
       "25694         tr_1007            .                       O     26290   \n",
       "25695         tr_1007        [SEP]                   [SEP]     26291   \n",
       "\n",
       "       New_Spacy_Tokens  \n",
       "25685  26281Chromosomal  \n",
       "25686     26282analysis  \n",
       "25687          26283was  \n",
       "25688       26284normal  \n",
       "25689            26285(  \n",
       "25690           2628646  \n",
       "25691            26287,  \n",
       "25692           26288XY  \n",
       "25693            26289)  \n",
       "25694            26290.  \n",
       "25695        26291[SEP]  "
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_token_level_annotated_data[final_token_level_annotated_data['New_Sentence_Id'] == 'tr_1007']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "dbfda2e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "token_tag_df_2['New_index'] = [every for every in range(len(token_tag_df_2))]\n",
    "token_tag_df_2['New_index'] = token_tag_df_2['New_index'].astype(str)\n",
    "token_tag_df_2['New_Spacy_Tokens'] = token_tag_df_2['New_index'] + token_tag_df_2['Token']\n",
    "token_tag_df_2.sort_values(by=['New_Sentence_Id','New_Spacy_Tokens'],inplace=True)\n",
    "token_tag_df_2.reset_index(drop=True,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "15cae1f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "      <th>New_index</th>\n",
       "      <th>New_Spacy_Tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25685</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>Chromosomal</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>205</td>\n",
       "      <td>205Chromosomal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25686</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>analysis</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>206</td>\n",
       "      <td>206analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25687</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>was</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>207</td>\n",
       "      <td>207was</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25688</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>normal</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>208</td>\n",
       "      <td>208normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25689</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>(</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>209</td>\n",
       "      <td>209(</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25690</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>46</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>210</td>\n",
       "      <td>21046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25691</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>211</td>\n",
       "      <td>211,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25692</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>XY</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>212</td>\n",
       "      <td>212XY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25693</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>)</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>213</td>\n",
       "      <td>213)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25694</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>.</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>214</td>\n",
       "      <td>214.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25695</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>215</td>\n",
       "      <td>215[SEP]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      New_Sentence_Id        Token Rules_Tag Rules_Tag_BIO New_index  \\\n",
       "25685         tr_1007  Chromosomal         O             O       205   \n",
       "25686         tr_1007     analysis         O             O       206   \n",
       "25687         tr_1007          was         O             O       207   \n",
       "25688         tr_1007       normal         O             O       208   \n",
       "25689         tr_1007            (         O             O       209   \n",
       "25690         tr_1007           46         O             O       210   \n",
       "25691         tr_1007            ,         O             O       211   \n",
       "25692         tr_1007           XY         O             O       212   \n",
       "25693         tr_1007            )         O             O       213   \n",
       "25694         tr_1007            .         O             O       214   \n",
       "25695         tr_1007        [SEP]     [SEP]         [SEP]       215   \n",
       "\n",
       "      New_Spacy_Tokens  \n",
       "25685   205Chromosomal  \n",
       "25686      206analysis  \n",
       "25687           207was  \n",
       "25688        208normal  \n",
       "25689             209(  \n",
       "25690            21046  \n",
       "25691             211,  \n",
       "25692            212XY  \n",
       "25693             213)  \n",
       "25694             214.  \n",
       "25695         215[SEP]  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "token_tag_df_2[token_tag_df_2['New_Sentence_Id'] == 'tr_1007']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "a9ec1bde",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_token_tag_df = pd.concat([token_tag_df_2[['New_Sentence_Id','Token','Rules_Tag_BIO']], \n",
    "                                final_token_level_annotated_data[['Human_Annotated_Tag_BIO']]],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "9c9cd8e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25685</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>Chromosomal</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25686</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>analysis</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25687</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>was</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25688</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>normal</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25689</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>(</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25690</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>46</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25691</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25692</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>XY</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25693</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>)</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25694</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>.</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25695</th>\n",
       "      <td>tr_1007</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      New_Sentence_Id        Token Rules_Tag_BIO Human_Annotated_Tag_BIO\n",
       "25685         tr_1007  Chromosomal             O                       O\n",
       "25686         tr_1007     analysis             O                       O\n",
       "25687         tr_1007          was             O                       O\n",
       "25688         tr_1007       normal             O                       O\n",
       "25689         tr_1007            (             O                       O\n",
       "25690         tr_1007           46             O                       O\n",
       "25691         tr_1007            ,             O                       O\n",
       "25692         tr_1007           XY             O                       O\n",
       "25693         tr_1007            )             O                       O\n",
       "25694         tr_1007            .             O                       O\n",
       "25695         tr_1007        [SEP]         [SEP]                   [SEP]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_token_tag_df[final_token_tag_df['New_Sentence_Id']== 'tr_1007']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "fb666b91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>Clustering</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>of</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>missense</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>mutations</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>in</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>the</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>ataxia</td>\n",
       "      <td>O</td>\n",
       "      <td>B-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>-</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>telangiectasia</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>gene</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>in</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>a</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>sporadic</td>\n",
       "      <td>O</td>\n",
       "      <td>B-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>T</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>-</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>cell</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>leukaemia</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>.</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>Ataxia</td>\n",
       "      <td>O</td>\n",
       "      <td>B-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>-</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>telangiectasia</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>(</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>A</td>\n",
       "      <td>O</td>\n",
       "      <td>B-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>-</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>T</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>)</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>is</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>a</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>recessive</td>\n",
       "      <td>O</td>\n",
       "      <td>B-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>multi</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>-</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>system</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>disorder</td>\n",
       "      <td>O</td>\n",
       "      <td>I-Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>caused</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>by</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>mutations</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>in</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>the</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>ATM</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>gene</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>at</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>11q22</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>-</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>q23</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>(</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>ref</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>.</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>3</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>)</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   New_Sentence_Id           Token Rules_Tag_BIO Human_Annotated_Tag_BIO\n",
       "0            tes_0      Clustering             O                       O\n",
       "1            tes_0              of             O                       O\n",
       "2            tes_0        missense             O                       O\n",
       "3            tes_0       mutations             O                       O\n",
       "4            tes_0              in             O                       O\n",
       "5            tes_0             the             O                       O\n",
       "6            tes_0          ataxia             O               B-Disease\n",
       "7            tes_0               -             O               I-Disease\n",
       "8            tes_0  telangiectasia             O               I-Disease\n",
       "9            tes_0            gene             O                       O\n",
       "10           tes_0              in             O                       O\n",
       "11           tes_0               a             O                       O\n",
       "12           tes_0        sporadic             O               B-Disease\n",
       "13           tes_0               T             O               I-Disease\n",
       "14           tes_0               -             O               I-Disease\n",
       "15           tes_0            cell             O               I-Disease\n",
       "16           tes_0       leukaemia             O               I-Disease\n",
       "17           tes_0               .             O                       O\n",
       "18           tes_0           [SEP]         [SEP]                   [SEP]\n",
       "19           tes_1          Ataxia             O               B-Disease\n",
       "20           tes_1               -             O               I-Disease\n",
       "21           tes_1  telangiectasia             O               I-Disease\n",
       "22           tes_1               (             O                       O\n",
       "23           tes_1               A             O               B-Disease\n",
       "24           tes_1               -             O               I-Disease\n",
       "25           tes_1               T             O               I-Disease\n",
       "26           tes_1               )             O                       O\n",
       "27           tes_1              is             O                       O\n",
       "28           tes_1               a             O                       O\n",
       "29           tes_1       recessive             O               B-Disease\n",
       "30           tes_1           multi             O               I-Disease\n",
       "31           tes_1               -             O               I-Disease\n",
       "32           tes_1          system             O               I-Disease\n",
       "33           tes_1        disorder             O               I-Disease\n",
       "34           tes_1          caused             O                       O\n",
       "35           tes_1              by             O                       O\n",
       "36           tes_1       mutations             O                       O\n",
       "37           tes_1              in             O                       O\n",
       "38           tes_1             the             O                       O\n",
       "39           tes_1             ATM             O                       O\n",
       "40           tes_1            gene             O                       O\n",
       "41           tes_1              at             O                       O\n",
       "42           tes_1           11q22             O                       O\n",
       "43           tes_1               -             O                       O\n",
       "44           tes_1             q23             O                       O\n",
       "45           tes_1               (             O                       O\n",
       "46           tes_1             ref             O                       O\n",
       "47           tes_1               .             O                       O\n",
       "48           tes_1               3             O                       O\n",
       "49           tes_1               )             O                       O"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_token_tag_df.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "42479792",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_token_tag_df['Human_Annotated_Tag_BIO'] = final_token_tag_df['Human_Annotated_Tag_BIO'].str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "5e90812f",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p ../data/diease_ner/output/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "73792b3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_token_tag_df.to_csv('../data/diease_ner/output/token_level_annotated_rules_data.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "324cad36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>191942</th>\n",
       "      <td>val_99</td>\n",
       "      <td>was</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191943</th>\n",
       "      <td>val_99</td>\n",
       "      <td>found</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191944</th>\n",
       "      <td>val_99</td>\n",
       "      <td>in</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191945</th>\n",
       "      <td>val_99</td>\n",
       "      <td>five</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191946</th>\n",
       "      <td>val_99</td>\n",
       "      <td>families</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191947</th>\n",
       "      <td>val_99</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191948</th>\n",
       "      <td>val_99</td>\n",
       "      <td>the</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191949</th>\n",
       "      <td>val_99</td>\n",
       "      <td>C</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191950</th>\n",
       "      <td>val_99</td>\n",
       "      <td>1806</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191951</th>\n",
       "      <td>val_99</td>\n",
       "      <td>T</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191952</th>\n",
       "      <td>val_99</td>\n",
       "      <td>nonsense</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191953</th>\n",
       "      <td>val_99</td>\n",
       "      <td>mutation</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191954</th>\n",
       "      <td>val_99</td>\n",
       "      <td>in</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191955</th>\n",
       "      <td>val_99</td>\n",
       "      <td>three</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191956</th>\n",
       "      <td>val_99</td>\n",
       "      <td>families</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191957</th>\n",
       "      <td>val_99</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191958</th>\n",
       "      <td>val_99</td>\n",
       "      <td>the</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191959</th>\n",
       "      <td>val_99</td>\n",
       "      <td>3166</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191960</th>\n",
       "      <td>val_99</td>\n",
       "      <td>insertion</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191961</th>\n",
       "      <td>val_99</td>\n",
       "      <td>TGAGA</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191962</th>\n",
       "      <td>val_99</td>\n",
       "      <td>in</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191963</th>\n",
       "      <td>val_99</td>\n",
       "      <td>three</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191964</th>\n",
       "      <td>val_99</td>\n",
       "      <td>families</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191965</th>\n",
       "      <td>val_99</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191966</th>\n",
       "      <td>val_99</td>\n",
       "      <td>and</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191967</th>\n",
       "      <td>val_99</td>\n",
       "      <td>the</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191968</th>\n",
       "      <td>val_99</td>\n",
       "      <td>nucleotide</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191969</th>\n",
       "      <td>val_99</td>\n",
       "      <td>1201</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191970</th>\n",
       "      <td>val_99</td>\n",
       "      <td>deletion</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191971</th>\n",
       "      <td>val_99</td>\n",
       "      <td>11</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191972</th>\n",
       "      <td>val_99</td>\n",
       "      <td>in</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191973</th>\n",
       "      <td>val_99</td>\n",
       "      <td>two</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191974</th>\n",
       "      <td>val_99</td>\n",
       "      <td>families</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191975</th>\n",
       "      <td>val_99</td>\n",
       "      <td>.</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191976</th>\n",
       "      <td>val_99</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       New_Sentence_Id       Token Rules_Tag_BIO Human_Annotated_Tag_BIO\n",
       "191942          val_99         was             O                       O\n",
       "191943          val_99       found             O                       O\n",
       "191944          val_99          in             O                       O\n",
       "191945          val_99        five             O                       O\n",
       "191946          val_99    families             O                       O\n",
       "191947          val_99           ,             O                       O\n",
       "191948          val_99         the             O                       O\n",
       "191949          val_99           C             O                       O\n",
       "191950          val_99        1806             O                       O\n",
       "191951          val_99           T             O                       O\n",
       "191952          val_99    nonsense             O                       O\n",
       "191953          val_99    mutation             O                       O\n",
       "191954          val_99          in             O                       O\n",
       "191955          val_99       three             O                       O\n",
       "191956          val_99    families             O                       O\n",
       "191957          val_99           ,             O                       O\n",
       "191958          val_99         the             O                       O\n",
       "191959          val_99        3166             O                       O\n",
       "191960          val_99   insertion             O                       O\n",
       "191961          val_99       TGAGA             O                       O\n",
       "191962          val_99          in             O                       O\n",
       "191963          val_99       three             O                       O\n",
       "191964          val_99    families             O                       O\n",
       "191965          val_99           ,             O                       O\n",
       "191966          val_99         and             O                       O\n",
       "191967          val_99         the             O                       O\n",
       "191968          val_99  nucleotide             O                       O\n",
       "191969          val_99        1201             O                       O\n",
       "191970          val_99    deletion             O                       O\n",
       "191971          val_99          11             O                       O\n",
       "191972          val_99          in             O                       O\n",
       "191973          val_99         two             O                       O\n",
       "191974          val_99    families             O                       O\n",
       "191975          val_99           .             O                       O\n",
       "191976          val_99       [SEP]         [SEP]                   [SEP]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_token_tag_df.tail(35)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "125d692b",
   "metadata": {},
   "source": [
    "### Aggregate the labels at sentence level"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a0b63f5",
   "metadata": {},
   "source": [
    "**Types of Sentences based on entities**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6d8ca14",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_token_tag_df_2 = final_token_tag_df.copy()\n",
    "\n",
    "final_token_tag_df_2 = final_token_tag_df_2[final_token_tag_df_2['Token'] != '[SEP]']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "d663259a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25588</th>\n",
       "      <td>tr_100</td>\n",
       "      <td>levels</td>\n",
       "      <td>O</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26509</th>\n",
       "      <td>tr_1035</td>\n",
       "      <td>ventilatory</td>\n",
       "      <td>O</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35490</th>\n",
       "      <td>tr_1339</td>\n",
       "      <td>be</td>\n",
       "      <td>O</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82607</th>\n",
       "      <td>tr_3</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103257</th>\n",
       "      <td>tr_37</td>\n",
       "      <td>activity</td>\n",
       "      <td>O</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107395</th>\n",
       "      <td>tr_3836</td>\n",
       "      <td>22</td>\n",
       "      <td>O</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107507</th>\n",
       "      <td>tr_384</td>\n",
       "      <td>less</td>\n",
       "      <td>O</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125489</th>\n",
       "      <td>tr_4481</td>\n",
       "      <td>to</td>\n",
       "      <td>O</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       New_Sentence_Id        Token Rules_Tag_BIO Human_Annotated_Tag_BIO\n",
       "25588           tr_100       levels             O                   [SEP]\n",
       "26509          tr_1035  ventilatory             O                   [SEP]\n",
       "35490          tr_1339           be             O                   [SEP]\n",
       "82607             tr_3            ,             O                   [SEP]\n",
       "103257           tr_37     activity             O                   [SEP]\n",
       "107395         tr_3836           22             O                   [SEP]\n",
       "107507          tr_384         less             O                   [SEP]\n",
       "125489         tr_4481           to             O                   [SEP]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_token_tag_df_2[final_token_tag_df_2['Human_Annotated_Tag_BIO']== '[SEP]']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "b8c440ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_token_tag_df_2['Human_Annotated_Tag_BIO'] = final_token_tag_df_2['Human_Annotated_Tag_BIO'].apply(lambda x: x if x!='[SEP]' else 'O')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "044c42cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [New_Sentence_Id, Token, Rules_Tag_BIO, Human_Annotated_Tag_BIO]\n",
       "Index: []"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_token_tag_df_2[final_token_tag_df_2['Human_Annotated_Tag_BIO']== '[SEP]']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "1fa932ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 146 ms, sys: 0 ns, total: 146 ms\n",
      "Wall time: 145 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sentence_level_tags</th>\n",
       "      <th>Count</th>\n",
       "      <th>%_contribution</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O</td>\n",
       "      <td>3337</td>\n",
       "      <td>46.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I-DISEASE|B-DISEASE|O</td>\n",
       "      <td>2698</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B-DISEASE|O</td>\n",
       "      <td>1260</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sentence_level_tags  Count  %_contribution\n",
       "0                      O   3337            46.0\n",
       "1  I-DISEASE|B-DISEASE|O   2698            37.0\n",
       "2            B-DISEASE|O   1260            17.0"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "sentence_level_df = final_token_tag_df_2.groupby('New_Sentence_Id')['Human_Annotated_Tag_BIO']\\\n",
    ".apply(lambda x: \"|\".join(\n",
    "    list(set(sorted(x)))\n",
    "))\n",
    "\n",
    "N = len(sentence_level_df)\n",
    "\n",
    "sentence_level_df = pd.DataFrame(sentence_level_df).reset_index()\n",
    "\n",
    "sentence_level_df = sentence_level_df[sentence_level_df['New_Sentence_Id'] != '[SEP]']\n",
    "\n",
    "sentence_level_df.reset_index(drop=True,inplace=True)\n",
    "\n",
    "sentence_level_count_df = pd.DataFrame(sentence_level_df['Human_Annotated_Tag_BIO'].value_counts()).reset_index()\n",
    "sentence_level_count_df.columns = ['Sentence_level_tags','Count']\n",
    "sentence_level_count_df['%_contribution'] = round(sentence_level_count_df['Count']/len(sentence_level_df) * 100.0)\n",
    "\n",
    "\n",
    "sentence_level_count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "8c08a7db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "O                              3333\n",
       "I-DISEASE|B-DISEASE|O          2695\n",
       "B-DISEASE|O                    1259\n",
       "[SEP]|O                           4\n",
       "I-DISEASE|B-DISEASE|[SEP]|O       3\n",
       "B-DISEASE|[SEP]|O                 1\n",
       "Name: Human_Annotated_Tag_BIO, dtype: int64"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence_level_df.Human_Annotated_Tag_BIO.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "009e30e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sentence_level_count_df.to_csv('spacy_model_ner/split_of_classes_sentence_level.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "7af06967",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7295, 2)"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence_level_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "291cdc26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>B-DISEASE|I-DISEASE|O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tes_1</td>\n",
       "      <td>B-DISEASE|I-DISEASE|O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tes_10</td>\n",
       "      <td>B-DISEASE|I-DISEASE|O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tes_100</td>\n",
       "      <td>B-DISEASE|I-DISEASE|O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tes_101</td>\n",
       "      <td>B-DISEASE|I-DISEASE|O</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  New_Sentence_Id Human_Annotated_Tag_BIO\n",
       "0           tes_0   B-DISEASE|I-DISEASE|O\n",
       "1           tes_1   B-DISEASE|I-DISEASE|O\n",
       "2          tes_10   B-DISEASE|I-DISEASE|O\n",
       "3         tes_100   B-DISEASE|I-DISEASE|O\n",
       "4         tes_101   B-DISEASE|I-DISEASE|O"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence_level_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a838066",
   "metadata": {},
   "source": [
    "Split the sentences into 80-10-10 train, dev and test datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "6c91b5cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "7a2af3d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df,intermediate_df =  train_test_split(sentence_level_df,\n",
    "                                             test_size=0.2, \n",
    "                                             stratify=sentence_level_df['Human_Annotated_Tag_BIO'],\n",
    "                                             random_state=0)\n",
    "\n",
    "\n",
    "dev_df, test_df = train_test_split(intermediate_df,\n",
    "                                             test_size=0.5, \n",
    "                                             stratify=intermediate_df['Human_Annotated_Tag_BIO'],\n",
    "                                             random_state=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "366a136c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5354</th>\n",
       "      <td>tr_4971</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>tes_267</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3577</th>\n",
       "      <td>tr_3371</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2366</th>\n",
       "      <td>tr_2281</td>\n",
       "      <td>I-DISEASE|B-DISEASE|O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>655</th>\n",
       "      <td>tes_689</td>\n",
       "      <td>B-DISEASE|O</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     New_Sentence_Id Human_Annotated_Tag_BIO\n",
       "5354         tr_4971                       O\n",
       "187          tes_267                       O\n",
       "3577         tr_3371                       O\n",
       "2366         tr_2281   I-DISEASE|B-DISEASE|O\n",
       "655          tes_689             B-DISEASE|O"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "f6500aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_token_level_data = final_token_tag_df[final_token_tag_df['New_Sentence_Id'].isin(train_df['New_Sentence_Id'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "bb5a7670",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(191977, 4)"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_token_tag_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "f52c9a77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(154023, 4)"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_token_level_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "e103d562",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p ../data/diease_ner/train_dev_test_split/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "2e8eb906",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_token_level_data.to_csv('../data/diease_ner/train_dev_test_split/train_token_level_df.csv', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "36314b8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>191972</th>\n",
       "      <td>val_99</td>\n",
       "      <td>in</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191973</th>\n",
       "      <td>val_99</td>\n",
       "      <td>two</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191974</th>\n",
       "      <td>val_99</td>\n",
       "      <td>families</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191975</th>\n",
       "      <td>val_99</td>\n",
       "      <td>.</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191976</th>\n",
       "      <td>val_99</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       New_Sentence_Id     Token Rules_Tag_BIO Human_Annotated_Tag_BIO\n",
       "191972          val_99        in             O                       O\n",
       "191973          val_99       two             O                       O\n",
       "191974          val_99  families             O                       O\n",
       "191975          val_99         .             O                       O\n",
       "191976          val_99     [SEP]         [SEP]                   [SEP]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_token_level_data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "f3b5da27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19203, 4)"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dev_token_level_data = final_token_tag_df[final_token_tag_df['New_Sentence_Id'].isin(dev_df['New_Sentence_Id'])]\n",
    "dev_token_level_data.to_csv('../data/diease_ner/train_dev_test_split/dev_token_level_df.csv', index=None)\n",
    "dev_token_level_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "d21b55c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18751, 4)"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_token_level_data = final_token_tag_df[final_token_tag_df['New_Sentence_Id'].isin(test_df['New_Sentence_Id'])]\n",
    "test_token_level_data.to_csv('../data/diease_ner/train_dev_test_split/test_token_level_df.csv', index=None)\n",
    "test_token_level_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "353c12b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_token_level_data.reset_index(drop=True,inplace=True)\n",
    "dev_token_level_data.reset_index(drop=True,inplace=True)\n",
    "test_token_level_data.reset_index(drop=True,inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f50c33bc",
   "metadata": {},
   "source": [
    "#### Spacy ML Compatible Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "f656eec5",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p ../data/diease_ner/train_dev_test_split_conll_data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "86e1b0a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_token_level_data = pd.read_csv('../data/diease_ner/train_dev_test_split/train_token_level_df.csv',index_col=False)\n",
    "dev_token_level_data = pd.read_csv('../data/diease_ner/train_dev_test_split/dev_token_level_df.csv',index_col=False)\n",
    "test_token_level_data = pd.read_csv('../data/diease_ner/train_dev_test_split/test_token_level_df.csv',index_col=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "74eb5113",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>Clustering</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>of</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>missense</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>mutations</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tes_0</td>\n",
       "      <td>in</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  New_Sentence_Id       Token Rules_Tag_BIO Human_Annotated_Tag_BIO\n",
       "0           tes_0  Clustering             O                       O\n",
       "1           tes_0          of             O                       O\n",
       "2           tes_0    missense             O                       O\n",
       "3           tes_0   mutations             O                       O\n",
       "4           tes_0          in             O                       O"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_token_level_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "38c501c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "O            136106\n",
       "I-DISEASE      6582\n",
       "[SEP]          5836\n",
       "B-DISEASE      5499\n",
       "Name: Human_Annotated_Tag_BIO, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_token_level_data['Human_Annotated_Tag_BIO'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "95cfd1b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_token_df_2_conll_string(token_tag_df,\n",
    "                                    token_column_name,\n",
    "                                    tag_column_name\n",
    "                                   ):\n",
    "    token_string = ''\n",
    "    for each in range(len(token_tag_df)):\n",
    "        if each %1000 == 0:\n",
    "            print(f\"{each} tokens processed\")\n",
    "        current_token_string = str(token_tag_df.loc[each,token_column_name])\n",
    "        current_tag_string = str(token_tag_df.loc[each,tag_column_name])\n",
    "        if current_tag_string == 'EP]':\n",
    "            current_tag_string = 'O'\n",
    "        if current_token_string !='[SEP]':\n",
    "            current_line = current_token_string + \"\\t\" + current_tag_string + \"\\n\"\n",
    "        else:\n",
    "            current_line = \"\\n\"\n",
    "        token_string = token_string + current_line\n",
    "    return token_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1e18bb19",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 tokens processed\n",
      "1000 tokens processed\n",
      "2000 tokens processed\n",
      "3000 tokens processed\n",
      "4000 tokens processed\n",
      "5000 tokens processed\n",
      "6000 tokens processed\n",
      "7000 tokens processed\n",
      "8000 tokens processed\n",
      "9000 tokens processed\n",
      "10000 tokens processed\n",
      "11000 tokens processed\n",
      "12000 tokens processed\n",
      "13000 tokens processed\n",
      "14000 tokens processed\n",
      "15000 tokens processed\n",
      "16000 tokens processed\n",
      "17000 tokens processed\n",
      "18000 tokens processed\n",
      "CPU times: user 324 ms, sys: 0 ns, total: 324 ms\n",
      "Wall time: 323 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "test_conll_string = convert_token_df_2_conll_string(test_token_level_data,\n",
    "                                'Token',\n",
    "                                'Human_Annotated_Tag_BIO'\n",
    "                               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "33ba68fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tes_102</td>\n",
       "      <td>Limits</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tes_102</td>\n",
       "      <td>to</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tes_102</td>\n",
       "      <td>accuracy</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tes_102</td>\n",
       "      <td>included</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tes_102</td>\n",
       "      <td>recombination</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  New_Sentence_Id          Token Rules_Tag_BIO Human_Annotated_Tag_BIO\n",
       "0         tes_102         Limits             O                       O\n",
       "1         tes_102             to             O                       O\n",
       "2         tes_102       accuracy             O                       O\n",
       "3         tes_102       included             O                       O\n",
       "4         tes_102  recombination             O                       O"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_token_level_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f8c00997",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/diease_ner/train_dev_test_split_conll_data/test_data.conll','w',encoding='utf-8') as f:\n",
    "    f.write(test_conll_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a277e470",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "investigate\tO\r\n",
      "the\tO\r\n",
      "rate\tO\r\n",
      "of\tO\r\n",
      "BRCA2\tO\r\n",
      "mutation\tO\r\n",
      "in\tO\r\n",
      "sporadic\tB-DISEASE\r\n",
      "breast\tI-DISEASE\r\n",
      "cancers\tI-DISEASE\r\n",
      "and\tO\r\n",
      "in\tO\r\n",
      "a\tO\r\n",
      "set\tO\r\n",
      "of\tO\r\n",
      "cell\tO\r\n",
      "lines\tO\r\n",
      "that\tO\r\n",
      "represent\tO\r\n",
      "twelve\tO\r\n",
      "other\tO\r\n",
      "tumour\tB-DISEASE\r\n",
      "types\tO\r\n",
      ".\tO\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!tail -n 25 ../data/diease_ner/train_dev_test_split_conll_data/test_data.conll"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "5cfedd2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18751 ../data/diease_ner/train_dev_test_split_conll_data/test_data.conll\r\n"
     ]
    }
   ],
   "source": [
    "!wc -l ../data/diease_ner/train_dev_test_split_conll_data/test_data.conll"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ca8f8993",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 tokens processed\n",
      "1000 tokens processed\n",
      "2000 tokens processed\n",
      "3000 tokens processed\n",
      "4000 tokens processed\n",
      "5000 tokens processed\n",
      "6000 tokens processed\n",
      "7000 tokens processed\n",
      "8000 tokens processed\n",
      "9000 tokens processed\n",
      "10000 tokens processed\n",
      "11000 tokens processed\n",
      "12000 tokens processed\n",
      "13000 tokens processed\n",
      "14000 tokens processed\n",
      "15000 tokens processed\n",
      "16000 tokens processed\n",
      "17000 tokens processed\n",
      "18000 tokens processed\n",
      "19000 tokens processed\n",
      "19203 ../data/diease_ner/train_dev_test_split_conll_data/dev_data.conll\n",
      "CPU times: user 341 ms, sys: 0 ns, total: 341 ms\n",
      "Wall time: 622 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "dev_conll_string = convert_token_df_2_conll_string(dev_token_level_data,\n",
    "                                'Token',\n",
    "                                'Human_Annotated_Tag_BIO'\n",
    "                               )\n",
    "\n",
    "\n",
    "with open('../data/diease_ner/train_dev_test_split_conll_data/dev_data.conll','w',encoding='utf-8') as f:\n",
    "    f.write(dev_conll_string)\n",
    "    \n",
    "!wc -l ../data/diease_ner/train_dev_test_split_conll_data/dev_data.conll"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a8e180c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 tokens processed\n",
      "1000 tokens processed\n",
      "2000 tokens processed\n",
      "3000 tokens processed\n",
      "4000 tokens processed\n",
      "5000 tokens processed\n",
      "6000 tokens processed\n",
      "7000 tokens processed\n",
      "8000 tokens processed\n",
      "9000 tokens processed\n",
      "10000 tokens processed\n",
      "11000 tokens processed\n",
      "12000 tokens processed\n",
      "13000 tokens processed\n",
      "14000 tokens processed\n",
      "15000 tokens processed\n",
      "16000 tokens processed\n",
      "17000 tokens processed\n",
      "18000 tokens processed\n",
      "19000 tokens processed\n",
      "20000 tokens processed\n",
      "21000 tokens processed\n",
      "22000 tokens processed\n",
      "23000 tokens processed\n",
      "24000 tokens processed\n",
      "25000 tokens processed\n",
      "26000 tokens processed\n",
      "27000 tokens processed\n",
      "28000 tokens processed\n",
      "29000 tokens processed\n",
      "30000 tokens processed\n",
      "31000 tokens processed\n",
      "32000 tokens processed\n",
      "33000 tokens processed\n",
      "34000 tokens processed\n",
      "35000 tokens processed\n",
      "36000 tokens processed\n",
      "37000 tokens processed\n",
      "38000 tokens processed\n",
      "39000 tokens processed\n",
      "40000 tokens processed\n",
      "41000 tokens processed\n",
      "42000 tokens processed\n",
      "43000 tokens processed\n",
      "44000 tokens processed\n",
      "45000 tokens processed\n",
      "46000 tokens processed\n",
      "47000 tokens processed\n",
      "48000 tokens processed\n",
      "49000 tokens processed\n",
      "50000 tokens processed\n",
      "51000 tokens processed\n",
      "52000 tokens processed\n",
      "53000 tokens processed\n",
      "54000 tokens processed\n",
      "55000 tokens processed\n",
      "56000 tokens processed\n",
      "57000 tokens processed\n",
      "58000 tokens processed\n",
      "59000 tokens processed\n",
      "60000 tokens processed\n",
      "61000 tokens processed\n",
      "62000 tokens processed\n",
      "63000 tokens processed\n",
      "64000 tokens processed\n",
      "65000 tokens processed\n",
      "66000 tokens processed\n",
      "67000 tokens processed\n",
      "68000 tokens processed\n",
      "69000 tokens processed\n",
      "70000 tokens processed\n",
      "71000 tokens processed\n",
      "72000 tokens processed\n",
      "73000 tokens processed\n",
      "74000 tokens processed\n",
      "75000 tokens processed\n",
      "76000 tokens processed\n",
      "77000 tokens processed\n",
      "78000 tokens processed\n",
      "79000 tokens processed\n",
      "80000 tokens processed\n",
      "81000 tokens processed\n",
      "82000 tokens processed\n",
      "83000 tokens processed\n",
      "84000 tokens processed\n",
      "85000 tokens processed\n",
      "86000 tokens processed\n",
      "87000 tokens processed\n",
      "88000 tokens processed\n",
      "89000 tokens processed\n",
      "90000 tokens processed\n",
      "91000 tokens processed\n",
      "92000 tokens processed\n",
      "93000 tokens processed\n",
      "94000 tokens processed\n",
      "95000 tokens processed\n",
      "96000 tokens processed\n",
      "97000 tokens processed\n",
      "98000 tokens processed\n",
      "99000 tokens processed\n",
      "100000 tokens processed\n",
      "101000 tokens processed\n",
      "102000 tokens processed\n",
      "103000 tokens processed\n",
      "104000 tokens processed\n",
      "105000 tokens processed\n",
      "106000 tokens processed\n",
      "107000 tokens processed\n",
      "108000 tokens processed\n",
      "109000 tokens processed\n",
      "110000 tokens processed\n",
      "111000 tokens processed\n",
      "112000 tokens processed\n",
      "113000 tokens processed\n",
      "114000 tokens processed\n",
      "115000 tokens processed\n",
      "116000 tokens processed\n",
      "117000 tokens processed\n",
      "118000 tokens processed\n",
      "119000 tokens processed\n",
      "120000 tokens processed\n",
      "121000 tokens processed\n",
      "122000 tokens processed\n",
      "123000 tokens processed\n",
      "124000 tokens processed\n",
      "125000 tokens processed\n",
      "126000 tokens processed\n",
      "127000 tokens processed\n",
      "128000 tokens processed\n",
      "129000 tokens processed\n",
      "130000 tokens processed\n",
      "131000 tokens processed\n",
      "132000 tokens processed\n",
      "133000 tokens processed\n",
      "134000 tokens processed\n",
      "135000 tokens processed\n",
      "136000 tokens processed\n",
      "137000 tokens processed\n",
      "138000 tokens processed\n",
      "139000 tokens processed\n",
      "140000 tokens processed\n",
      "141000 tokens processed\n",
      "142000 tokens processed\n",
      "143000 tokens processed\n",
      "144000 tokens processed\n",
      "145000 tokens processed\n",
      "146000 tokens processed\n",
      "147000 tokens processed\n",
      "148000 tokens processed\n",
      "149000 tokens processed\n",
      "150000 tokens processed\n",
      "151000 tokens processed\n",
      "152000 tokens processed\n",
      "153000 tokens processed\n",
      "154000 tokens processed\n",
      "154023 ../data/diease_ner/train_dev_test_split_conll_data/train_data.conll\n",
      "CPU times: user 2.66 s, sys: 0 ns, total: 2.66 s\n",
      "Wall time: 2.94 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "train_conll_string = convert_token_df_2_conll_string(train_token_level_data,\n",
    "                                'Token',\n",
    "                                'Human_Annotated_Tag_BIO'\n",
    "                               )\n",
    "\n",
    "\n",
    "with open('../data/diease_ner/train_dev_test_split_conll_data/train_data.conll','w',encoding='utf-8') as f:\n",
    "    f.write(train_conll_string)\n",
    "    \n",
    "!wc -l ../data/diease_ner/train_dev_test_split_conll_data/train_data.conll"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "49854ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_of_tags(df,\n",
    "                  name_of_file\n",
    "                 ):\n",
    "    df = pd.DataFrame(df['Human_Annotated_Tag_BIO'].value_counts()).reset_index()\n",
    "    df.columns = ['Sentence_level_tags','Count']\n",
    "    df['%_contribution'] = round(df['Count']/sum(df['Count']) * 100.0)\n",
    "    df.to_csv(f'spacy_model_ner/{name_of_file}',index=None)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "5d07e514",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_count_df = count_of_tags(train_df,name_of_file='split_of_classes_sentence_level_train.csv')\n",
    "dev_count_df = count_of_tags(dev_df,name_of_file='split_of_classes_sentence_level_dev.csv')\n",
    "test_count_df = count_of_tags(test_df,name_of_file='split_of_classes_sentence_level_test.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "9867c709",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sentence_level_tags</th>\n",
       "      <th>Count</th>\n",
       "      <th>%_contribution</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O</td>\n",
       "      <td>2670</td>\n",
       "      <td>46.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I-DISEASE|B-DISEASE|O</td>\n",
       "      <td>2158</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B-DISEASE|O</td>\n",
       "      <td>1008</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sentence_level_tags  Count  %_contribution\n",
       "0                      O   2670            46.0\n",
       "1  I-DISEASE|B-DISEASE|O   2158            37.0\n",
       "2            B-DISEASE|O   1008            17.0"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "c9fb0960",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sentence_level_tags</th>\n",
       "      <th>Count</th>\n",
       "      <th>%_contribution</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O</td>\n",
       "      <td>333</td>\n",
       "      <td>46.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I-DISEASE|B-DISEASE|O</td>\n",
       "      <td>270</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B-DISEASE|O</td>\n",
       "      <td>126</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sentence_level_tags  Count  %_contribution\n",
       "0                      O    333            46.0\n",
       "1  I-DISEASE|B-DISEASE|O    270            37.0\n",
       "2            B-DISEASE|O    126            17.0"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dev_count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "733fe397",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sentence_level_tags</th>\n",
       "      <th>Count</th>\n",
       "      <th>%_contribution</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O</td>\n",
       "      <td>334</td>\n",
       "      <td>46.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I-DISEASE|B-DISEASE|O</td>\n",
       "      <td>270</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B-DISEASE|O</td>\n",
       "      <td>126</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sentence_level_tags  Count  %_contribution\n",
       "0                      O    334            46.0\n",
       "1  I-DISEASE|B-DISEASE|O    270            37.0\n",
       "2            B-DISEASE|O    126            17.0"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_count_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cce4b23",
   "metadata": {},
   "source": [
    "### Appendix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "5d97b323",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = token_tag_df_2[token_tag_df_2['New_Sentence_Id']=='tr_0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "e3bda4d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df.to_csv('spacy_model_ner/token_level_tags_on_one_unlabeled_sentence.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "7038ceaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df_2 = final_token_tag_df[final_token_tag_df['New_Sentence_Id']=='tr_0']\n",
    "new_df_2.to_csv('spacy_model_ner/token_level_tags_on_one_unlabeled_sentence_2.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "12115b63",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>New_Sentence_Id</th>\n",
       "      <th>Token</th>\n",
       "      <th>Rules_Tag_BIO</th>\n",
       "      <th>Human_Annotated_Tag_BIO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>Identification</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>of</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>of</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>of</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>of</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>APC2</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>a</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>homologue</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>the</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>adenomatous</td>\n",
       "      <td>B-DISEASE</td>\n",
       "      <td>B-DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>polyposis</td>\n",
       "      <td>I-DISEASE</td>\n",
       "      <td>I-DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>coli</td>\n",
       "      <td>I-DISEASE</td>\n",
       "      <td>I-DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>tumour</td>\n",
       "      <td>O</td>\n",
       "      <td>I-DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>suppressor</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>.</td>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>tr_0</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "      <td>[SEP]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   New_Sentence_Id           Token Rules_Tag_BIO Human_Annotated_Tag_BIO\n",
       "0             tr_0  Identification             O                       O\n",
       "1             tr_0              of             O                       O\n",
       "2             tr_0              of             O                       O\n",
       "3             tr_0              of             O                       O\n",
       "4             tr_0              of             O                       O\n",
       "5             tr_0            APC2             O                       O\n",
       "6             tr_0               ,             O                       O\n",
       "7             tr_0               a             O                       O\n",
       "8             tr_0       homologue             O                       O\n",
       "9             tr_0             the             O                       O\n",
       "10            tr_0     adenomatous     B-DISEASE               B-DISEASE\n",
       "11            tr_0       polyposis     I-DISEASE               I-DISEASE\n",
       "12            tr_0            coli     I-DISEASE               I-DISEASE\n",
       "13            tr_0          tumour             O               I-DISEASE\n",
       "14            tr_0      suppressor             O                       O\n",
       "15            tr_0               .             O                       O\n",
       "16            tr_0           [SEP]         [SEP]                   [SEP]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df_2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
